### ISSUE TITLE
weights for the movenet_mobilenetv3.py

### ISSUE THREAD

**Issue by mukeshnarendran7** [fire717/movenet.pytorch#19]
Hi, I am not able to match the pre-trained weights for the above model file from the output path. Could you guide me on how i can use the v3 model. Thanks

**Comment by fire717** [comment]
v3 is just for test and it's acc is lower than v2 in my test, also origin movenet backbone is v2, so if u wanna use this I suggest v2.
If u still wanna try v3, u can change 
https://github.com/fire717/movenet.pytorch/blob/master/lib/__init__.py#L7-L9

```
from lib.models.movenet_mobilenetv2 import MoveNet
```
to
```
from lib.models.movenet_mobilenetv3 import MoveNet
```

**Comment by mukeshnarendran7** [comment]
Thanks for getting back. I have another question regarding the inputs and outputs.

- So the inputs are just : (1,3, 192, 192) or something more
- outputs have a list of 4 values. I have used the index 0 for plotting heatmaps similar to the below. Is this the right way to do it? What is the remainder then of the outputs used for? They seem like heatmaps itself


def get_keypoints(heatmaps, thr=0.5):
    n, h, w = heatmaps.size()
    flat = heatmaps.view(n, -1)
    max_val, max_idx = flat.max(dim=1)
    xx = (max_idx % w).view(-1, 1)  # (-1,1) for column vector
    yy = (max_idx // w).view(-1, 1)  # (-1,1) for column vector
    xx[max_val <= thr] = -1
    yy[max_val <= thr] = -1
    keypoints = torch.cat((xx, yy), dim=1)
    keypoints = keypoints.numpy()
    #re-scale them back
    x =  keypoints[:,0]*(640/48)
    y =  keypoints[:,1]*(640/48)
    return x, y

**Comment by fire717** [comment]
1. As for this repo, inputs are just : (1,3, 192, 192), its the MoveNet Lightning version . If u wanna use other size, u need to change the weight matrix, model head,etc.
2. Refer to [Official blog](https://blog.tensorflow.org/2021/05/next-generation-pose-detection-with-movenet-and-tensorflowjs.html) "MoveNet Architecture" part.
3. Of course u can try any loss, but its not a comparison with  bone loss, "this is a muti-task learning" in readme.
4. The difference is that they are different model.
5. Just as the code default setting.
6. In my test, post traing quantize is not helpful, QAT may help, but I cannot find a easy way to use QAT in PyTorch(Google use Tensorflow.)

Most of ur question can be found in readme or source code, just dive it if u ara interested in MoveNet!

**Comment by mukeshnarendran7** [comment]
Hey fire717, Thanks for the reply.  How can I fine-tune it for say 16 key points instead of 17? When I change the header to match 17 key points then it looks like the model trains from scratch and then overfits. A code example would be nice. Thanks

**Comment by fire717** [comment]
fine-tune it for 16 key points is a little complecated , cause the target of this repo is to reproduce origin movenet, if u wanna use different size or points numbers, u may read the article and the source code and understand them then change the code for urself.
kps_mask is use to filter unseen points to avoid computing loss
bone IDX is not important, u can try some different set to test

**Comment by mukeshnarendran7** [comment]
Hey, I tried to introduce a dummy column with zero values to match the key points from 16 ->17 and set it as not labelled.

1. I have some issues with inputs: My format looks like this for one item. I am not sure how to set the other_keypoints and other_centers. How do you get them because my images have only one object per frame and should I set them to -1?
2. **class TensorDataset(Dataset)**: I am using the is data loader and passing these values to get the heatmaps (via **label2heatmap**) and they return me zeros(48, 48). Any hints on what I could be missing from the item. I saw that my area is extremely large (-in 1000's)
Thanks

```
Keypoints: [167.09727626459144, 41.119730185497474, 2, 147.1750972762646, 53.74704890387858, 2, 116.04669260700388, 37.881956155143335, 2, 108.32684824902724, 88.06745362563237, 2, 120.77821011673151, 123.35919055649242, 2, 104.34241245136187, 142.4620573355818, 2, 129.4941634241245, 88.39123102866779, 2, 135.71984435797668, 121.09274873524453, 2, 148.42023346303503, 159.62225969645868, 2, 38.84824902723736, 32.377740303541316, 2, 38.59922178988327, 80.94435075885329, 2, 33.369649805447466, 109.11298482293424, 2, 55.53307392996109, 143.43338954468803, 2, 29.883268482490273, 79.97301854974705, 2, 24.902723735408564, 106.1989881956155, 2, 35.36186770428016, 149.58516020236087, 2, 0.0, 0.0, 0]
center: [96, 96]
other_keypoints: [[167.09727626459144, 41.119730185497474], [147.1750972762646, 53.74704890387858], [116.04669260700388, 37.881956155143335], [108.32684824902724, 88.06745362563237], [120.77821011673151, 123.35919055649242], [104.34241245136187, 142.4620573355818], [129.4941634241245, 88.39123102866779], [135.71984435797668, 121.09274873524453], [148.42023346303503, 159.62225969645868], [38.84824902723736, 32.377740303541316], [38.59922178988327, 80.94435075885329], [33.369649805447466, 109.11298482293424], [55.53307392996109, 143.43338954468803], [29.883268482490273, 79.97301854974705], [24.902723735408564, 106.1989881956155], [35.36186770428016, 149.58516020236087], [0.0, 0.0]]
other_center: [[-1, -1]]
```

**Comment by fire717** [comment]
If u have only one object, other_keypoints should be {}
Yes, key points value is relative value of x,  such as 0.64,  absolute value = 0.64x192

As u have so many questions, I think u  maybe didnot know how the MoveNet work, maybe u should read the bolg firstly, then read the source code of this repo and understand it by urself.

**Comment by mukeshnarendran7** [comment]
Thanks for sharing the article. I read through it and was helpful. When i pass in my data for preparation using your code label2center. My centres look like below with negative values but are positioned well. The max and min values are large. I am scaling them down like the key points. My heatmaps look okay though. Any hints on what I could be doing wrong. Thanks


       ```
        center = [(192//2)/192, (192//2)/192]
        heatmaps,sigma = label2heatmap(keypoints, other_keypoints, self.img_size)
        cx = min(max(0,int(center[0]*self.img_size//4)),self.img_size//4-1)
        cy = min(max(0,int(center[1]*self.img_size//4)),self.img_size//4-1)
        
        #Denter heatmaps
        centers = label2center(cx=cx, cy=cy, other_centers=other_centers, img_size=192, sigma=sigma)#(1, 48, 48)

        regs = label2reg(keypoints, cx, cy, self.img_size) #(14, 48, 48)
        offsets = label2offset(keypoints, cx, cy, regs, self.img_size)#(14, 48, 48)
        labels = np.concatenate([heatmaps,centers,regs,offsets],axis=0)
```
![image](https://user-images.githubusercontent.com/23098703/163848855-93fedaef-93eb-47d5-8221-2a7090debc88.png)
![image](https://user-images.githubusercontent.com/23098703/163849748-39b85836-6f6c-4fd9-b31d-3a6a30352cb7.png)

Also, my loss values are quite large but the model seems to be learning:
![image](https://user-images.githubusercontent.com/23098703/163849276-c30a33f6-7b3f-4cee-94a0-3e3ed4e4e2cf.png)

**Comment by fire717** [comment]
label2center  just  convert  ur  center point location (x,y) to a heatmap feature map, so if it's not right, maybe because the (x,y) for your center is not right.
The center point is calculate in https://github.com/fire717/movenet.pytorch/blob/master/scripts/make_coco_data_17keypooints.py

which meaning is the center point of all keypoints of one object.
